src/sfloader/fluid_sffile.c

```diff
@@ -1516,11 +1516,12 @@ static int load_pgen(SFData *sf, int size)
                 }
                 else
                 {
+                    SFZone * pzone = fluid_list_get(p2);
                     /* previous global zone exists, discard */
                     FLUID_LOG(FLUID_WARN, "Preset '%s': Discarding invalid global zone",
                               ((SFPreset *)(p->data))->name);
-                    *hz = fluid_list_remove(*hz, p2->data);
-                    delete_zone((SFZone *)fluid_list_get(p2));
+                    *hz = fluid_list_remove(start_of_zone_list, pzone);
```

Comment:
I meant this line with the previous comment. Got confused with the side-by-side diff view...

Reply:
I think simply removing the assignment to `*hz` would be ok here. Same in `load_iggen()`.

Reply:
You're right. I think it should have been

```suggestion
                    start_of_zone = fluid_list_remove(start_of_zone_list, pzone);
```

And then we have to save `p2->next`, advance `p2`, and `continue` the loop. Also, `p3` in the `while(p3)` statement below could still point to the zone we've just deleted, couldn't it?

Reply:
Nope, we cannot remove the assignment. We need it in case we are dealing with the first zone to adjust the beginning of the list.

Reply:
We will never deal with the first zone here, because the prerequisite ist that there is a previous global zone.

Reply:
Ofc, sry.

Reply:
But you are right, need also need to continue here. So we should be safe to do the following:
```
fluid_list_remove(start_of_zone_list, pzone);
delete_zone(pzone);
p2 = fluid_list_next(p2);
continue; /* zone iteration */
```